Day 16: “Controlling dialogs and popovers with the Invoker Commands API” / Aubrey Sambor

starshaped opened the pull request:
This is ready for feedback ❤️

Commit:
Initial commit of day 16 post

Commit:
Update title and make tweaks

Commit:
More tweaks

Commit:
One more tweak

matuzo commented on hell/adventcalendar/2025/16/index.md:
```diff
@@ -1,35 +1,114 @@
 
-## Heading
+* Dialogs usually contain an action to take, such as agreeing to text or choosing an option. Popovers usually display short-lived information such as a date picker or toast notification.
+* A dialog must be explicitly closed, either by taking an action within the dialog or closing via a close button, while a popover can be lightly dismissed by clicking outside of the popover.
```
That's not entirely true anymore because dialogs now also have light dismiss.

https://html.spec.whatwg.org/multipage/interactive-elements.html#dialog-light-dismiss

hidde replied:
yeah… true. Maybe adding “by default” helps? As with dialogs it isn't a default behaviour, it needs `closedby`.

starshaped replied:
Went ahead and added 'by default' to the sentence, thanks for pointing out that dialogs now also have a light dismiss option!

matuzo commented on hell/adventcalendar/2025/16/index.md:
```diff
@@ -1,35 +1,114 @@
+* A dialog must be explicitly closed, either by taking an action within the dialog or closing via a close button, while a popover can be lightly dismissed by clicking outside of the popover.
+* A dialog can either be modal or non-modal, while a popover is always non-modal by design. As of late 2025 the Invoker Commands API only has a built-in command for a modal dialog, so I will be focusing on modal dialogs only in this post.
+* Dialogs have a built in backdrop functionality to tint the background a different color, while popovers shouldn't use a backdrop in most cases.
+* A dialog is an HTML element with a role of `dialog`, while a popover is considered an attribute and does not come with a role.
```
Yes and no. It doesn't come with a role but if you put popover on a generic element like div, Chromium-based browsers and Firefox turn the generic div into a group, as if you where using role=group explicitly.

hidde replied:
Agree, that would be good nuance to add.

I do think it's helpful to frame popover as 'needs developer to consider what the role should be'. You could link out to my post on [popover roles](https://hidde.blog/popover-semantics/)?

starshaped replied:
Went ahead and added more info about popover roles and added a link to your post, Hidde!

matuzo reviewed (commented):
Great work, thank you!

hidde commented on hell/adventcalendar/2025/16/index.md:
```diff
@@ -1,35 +1,114 @@
 ---
-Some text.
-Some text.
+The [Invoker Commands API](https://developer.mozilla.org/en-US/docs/Web/API/Invoker_Commands_API) is a new way to control popovers and modal dialogs without having to write JavaScript. Dialogs have been available in all modern browsers since March 2022, and the Popover API is available in all modern browsers as of January 2025.
```
wonder if it's worth introducing Invoker Commands as a way to assign behaviours to buttons, rather than a thing just for popovers and dialogs?

starshaped replied:
Yeah, I like this idea better as that's what it is, new button behaviors. I'll retool this to frame it this way. Thanks!

starshaped replied:
Rewrote the intro paragraph slightly to focus on the API as button behaviors.

hidde commented on hell/adventcalendar/2025/16/index.md:
```diff
@@ -1,35 +1,114 @@
+* `show-popover`: Shows a popover element. This is equivalent to both the `popoveraction="show"` attribute and the `HTMLElement.showPopover()` method.
+* `hide-popover`: Hides a popover element. This is equivalent to both the `popoveraction="hide"` attribute and the `HTMLElement.hidePopover()` method.
+* `toggle-popover`: Toggles a popover element; if the popover is hidden, it will be shown, and if the popover is shown, it will be hidden. This is equivalent to both the `popoveraction="toggle"` attribute and the `HTMLElement.togglePopover()` method.
+* Custom values: This is where The Invoker API really shines! A user can provide a custom value, prefixed by two hyphens (`--`) to create a custom `command` event. This event is usually written in JavaScript and provided by the user to add new functionality. The possibilities are endless!
```
later in the post you mention some examples, could add a 'more on that later' forward reference here?

starshaped replied:
I have the examples section right after this list of attributes, do you mean I should specifically call out the custom values example or add a 'more on that later' below this?

hidde replied:
ah, you're right the examples are right after; reading it again I'd say this works fine as is!

hidde reviewed (commented):
Great post! Added a few suggestions.

Commit:
Add by default to the light dismiss section of dialogs

Commit:
Add clarification about popover roles

Commit:
Update intro to focus on button functionality

starshaped commented:
Thanks, @matuzo and @hidde for your feedback on my article! I've addressed most of the suggestions and asked a question on the other. This was a joy to write and so much fun!

Commit:
Tweak intro

hidde commented:
It was a lovely read, Aubrey, like how it turned out!

matuzo commented:
Fantastic, thank you. Are you happy with the article? Can I merge it @starshaped?

starshaped commented:
@matuzo Yes, I'm very happy with the article and thanks again for your feedback! Feel free to merge it :)

matuzo commented:
Perfect! :)
